### PR TITLE
Add public API MemoryMarshal.GetArrayDataReference

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Resolves https://github.com/dotnet/corefx/issues/36133.

This introduces a new public API `public static MemoryMarshal.GetRawArrayData<T>(T[] array) : ref T`. It's similar in spirit to the existing `MemoryMarshal.GetReference`, with the following explicit behaviors:

 * The method returns a reference to the first element in the array; or, if the input array is zero-length, a reference to where the first element would have been. This reference can be used for address comparisons (such as `Unsafe.IsAddressGreaterThan`) in all cases, but it can only be safely dereferenced in the case where the input array is _not_ zero-length.
 * No argument null check is performed. If a null reference is provided, the method throws NRE. Contrast this against `MemoryMarshal.GetReference`, which will return _nullptr_ (as a _T\&_) when given a `Span<T>` generated from a null `T[]`.
 * No array variance check is performed. It's possible to pass a `string[]` to this API and to get a _ref object_ back. Such a thing can be safely dereferenced (subject to the zero-length argument caveat mentioned above), but it must never be written to unless the caller knows that the value being written is compatible with the actual type of the array.

For an example of that last bullet point above:

```cs
/* using standard "safe" ref semantics */
string[] strArray = new string[] { "Hello!" };
ref string s = ref strArray[0]; // succeeds
object[] objArray = strArray; // succeeds, actual type is string[]
ref object o = ref array[0]; // throws ArrayTypeMismatchException due to runtime array variance checks

/* using "unsafe" GetRawArrayData */
object[] array = new string[] { "Hello!" }; // actual type is string[]
ref object o = ref MemoryMarshal.GetRawArrayData<object>(array); // succeeds
Console.WriteLine(o); // dereferencing 'o' is OK as long as array.Length >= 1
o = new object(); // !! DON'T DO THIS !! - this is a type safety violation
```

In API review, we had decided on naming this method `MemoryMarshal.GetReference` to match the existing span-based overloads. However, after some experimentation I realized that the logic we based our decision on was faulty - see my comment at https://github.com/dotnet/corefx/issues/36133#issuecomment-566834006 for more info. So for this PR I've reverted back to the originally suggested name. If we want to keep the `GetReference` name anyway I can do that - just let me know and I can update this PR.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Memory.cs [hunk 1]
```diff
@@ -321,13 +321,17 @@ public unsafe Span<T> Span
                     {
                         // We know the object is not null, it's not a string, and it is variable-length. The only
                         // remaining option is for it to be a T[] (or a U[] which is blittable to T[], like int[]
-                        // and uint[]). Otherwise somebody used private reflection to set this field, and we're not
-                        // too worried about type safety violations at this point.
+                        // and uint[]). As a special case of the this, ROM<T> allows some amount of array variance
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Private.CoreLib/src/System/Memory.cs]: _the this_ is a typo - will fix.

File: src/libraries/System.Private.CoreLib/src/System/ReadOnlyMemory.cs [hunk 2]
```diff
@@ -243,13 +243,17 @@ public unsafe ReadOnlySpan<T> Span
                     {
                         // We know the object is not null, it's not a string, and it is variable-length. The only
                         // remaining option is for it to be a T[] (or a U[] which is blittable to T[], like int[]
-                        // and uint[]). Otherwise somebody used private reflection to set this field, and we're not
-                        // too worried about type safety violations at this point.
+                        // and uint[]). As a special case of the this, ROM<T> allows some amount of array variance
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Private.CoreLib/src/System/ReadOnlyMemory.cs]: _the this_ is a typo - will fix.

File: src/libraries/System.Private.CoreLib/src/System/ReadOnlyMemory.cs [hunk 3]
```diff
@@ -358,13 +362,13 @@ public unsafe MemoryHandle Pin()
                     // Array is already pre-pinned
                     if (_index < 0)
                     {
-                        void* pointer = Unsafe.Add<T>(Unsafe.AsPointer(ref Unsafe.As<T[]>(tmpObject).GetRawSzArrayData()), _index & RemoveFlagsBitMask);
+                        void* pointer = Unsafe.Add<T>(Unsafe.AsPointer(ref MemoryMarshal.GetRawArrayData(Unsafe.As<T[]>(tmpObject))), _index & ReadOnlyMemory<T>.RemoveFlagsBitMask);
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Private.CoreLib/src/System/ReadOnlyMemory.cs]: Will revert `ReadOnlyMemory<T>.RemoveFlagsBitMask` back to simply `RemoveFlagsBitMask`. Missed this during initial review.

File: src/coreclr/src/System.Private.CoreLib/src/System/Array.CoreCLR.cs [hunk 4]
```diff
@@ -150,8 +151,8 @@ public static unsafe void Copy(Array sourceArray, Array destinationArray, int le
                 (uint)length <= (nuint)destinationArray.LongLength)
             {
                 nuint byteCount = (uint)length * (nuint)pMT->ComponentSize;
-                ref byte src = ref sourceArray.GetRawSzArrayData();
-                ref byte dst = ref destinationArray.GetRawSzArrayData();
+                ref byte src = ref MemoryMarshal.GetRawArrayData(Unsafe.As<byte[]>(sourceArray));
```
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/src/System/Array.CoreCLR.cs]: The array is not an actual `byte[]` array. It is a confusing to unsafe cast the array to `byte[]` when it is not actually `byte[]`

File: src/coreclr/src/vm/jitinterface.cpp [hunk 5]
```diff
@@ -7243,6 +7243,39 @@ bool getILIntrinsicImplementationForUnsafe(MethodDesc * ftn,
     return false;
 }
 
+bool getILIntrinsicImplementationForMemoryMarshal(MethodDesc * ftn,
```
Comment (jkotas) [inline src/coreclr/src/vm/jitinterface.cpp]: Could you please the intrinsic expansion to `src\coreclr\src\tools\Common\TypeSystem\IL\Stubs` as well?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs [hunk 6]
```diff
@@ -170,7 +170,7 @@ public static unsafe IntPtr UnsafeAddrOfPinnedArrayElement<T>(T[] arr, int index
             if (arr is null)
                 throw new ArgumentNullException(nameof(arr));
 
-            void* pRawData = Unsafe.AsPointer(ref arr.GetRawSzArrayData());
+            void* pRawData = Unsafe.AsPointer(ref MemoryMarshal.GetRawArrayData(arr));
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs]: This change will make the API slower.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs [hunk 7]
```diff
@@ -76,6 +76,19 @@ public static ReadOnlySpan<byte> AsBytes<T>(ReadOnlySpan<T> span)
...
+        /// <summary>
+        /// Returns a reference to the 0th element of <paramref name="array"/>. If the array is empty, returns a reference to where the 0th element
+        /// would have been stored. Such a reference may be used for pinning but must never be dereferenced.
+        /// </summary>
+        /// <exception cref="NullReferenceException">If <paramref name="array"/> is <see langword="null"/>.</exception>
+        /// <remarks>
+        /// This method does not perform array variance checks. The caller must manually perform any array variance checks
+        /// if the caller wishes to write to the returned reference.
+        /// </remarks>
+        [Intrinsic]
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: We may want to mark it `NonVersionable`, so it gets inlined across version bubbles.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: (The native images assume array layout, so we are not losing any versionability by inlining this method.)

File: src/coreclr/src/System.Private.CoreLib/src/System/Array.CoreCLR.cs (hunk 4 above)
Comment (jkotas) [reply src/coreclr/src/System.Private.CoreLib/src/System/Array.CoreCLR.cs]: Add a comment at least.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs [hunk 8]
```diff
@@ -76,6 +76,19 @@ public static ReadOnlySpan<byte> AsBytes<T>(ReadOnlySpan<T> span)
...
+        public static ref T GetRawArrayData<T>(T[] array) =>
+            ref Unsafe.As<byte, T>(ref Unsafe.As<RawArrayData>(array).Data);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: This should be in the runtime specific portion, like all other references to `RawArrayData`.

File: src/coreclr/src/System.Private.CoreLib/src/System/Array.CoreCLR.cs (hunk 4 above)
Comment (GrabYourPitchforks) [reply src/coreclr/src/System.Private.CoreLib/src/System/Array.CoreCLR.cs]: Would you prefer something like `Unsafe.As<T, byte>(ref GetRawArrayData(...))` instead?

Comment (jkotas) [reply src/coreclr/src/System.Private.CoreLib/src/System/Array.CoreCLR.cs]: > `Unsafe.As<T, byte>(ref GetRawArrayData(...))`

I am not sure how it would here. There is no `T` available here.

The fact that layout is same for all SZ arrays was used as an implementation detail of GetRawArrayData before this change. Now, it is leaking out silently here.

What about changing this to `ref byte src = ref Unsafe.As<RawArrayData>(sourceArray);`? It would make me happy because it makes it obvious that it is tightly coupled with array layout.

**Review by jkotas — APPROVED** [review]
LGTM (modulo the decision on the name).

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.CoreCLR.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace System.Runtime.InteropServices
+{
+    public static partial class MemoryMarshal
+    {
+        /// <summary>
+        /// Returns a reference to the 0th element of <paramref name="array"/>. If the array is empty, returns a reference to where the 0th element
+        /// would have been stored. Such a reference may be used for pinning but must never be dereferenced.
+        /// </summary>
+        /// <exception cref="NullReferenceException">If <paramref name="array"/> is <see langword="null"/>.</exception>
```
Comment (ahsonkhan) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.CoreCLR.cs]: doc nit:
```suggestion
        /// <exception cref="NullReferenceException"><paramref name="array"/> is <see langword="null"/>.</exception>
```

https://github.com/dotnet/dotnet-api-docs/wiki/Exceptions#how-to
> Write each condition as if the word "if" precedes it.

File: src/libraries/System.Memory/tests/MemoryMarshal/GetRawArrayData.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        public static void GetRawArrayData_NonEmptyInput_ReturnsRefToFirstElement()
+        {
+            int[] theArray = new int[] { 10, 20, 30 };
+            Assert.True(Unsafe.AreSame(ref theArray[0], ref MemoryMarshal.GetRawArrayData(theArray)));
+        }
+
+        [Fact]
+        public static unsafe void GetRawArrayData_EmptyInput_ReturnsRefToWhereFirstElementWouldBe()
+        {
+            int[] theArray = new int[0];
```
Comment (ahsonkhan) [inline src/libraries/System.Memory/tests/MemoryMarshal/GetRawArrayData.cs]: You mentioned above in the API doc:
> Such a reference may be used for pinning but must never be dereferenced.

What will happen if you did deference it?

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 11]
```diff
@@ -1115,31 +1116,31 @@ public static int IndexOf<T>(T[] array, T value, int startIndex, int count)
                 if (Unsafe.SizeOf<T>() == sizeof(byte))
                 {
                     int result = SpanHelpers.IndexOf(
-                        ref Unsafe.Add(ref array.GetRawSzArrayData(), startIndex),
+                        ref Unsafe.Add(ref MemoryMarshal.GetRawArrayData(Unsafe.As<byte[]>(array)), startIndex),
```
Comment (ahsonkhan) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs]: Can you explain why this change is better? Is the goal to avoid using `GetRawSzArrayData`?

**Review by ahsonkhan — APPROVED** [review]

File: src/libraries/System.Memory/tests/MemoryMarshal/GetRawArrayData.cs (hunk 10 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Memory/tests/MemoryMarshal/GetRawArrayData.cs]: The behavior is undefined. It will most likely return a nonsensical value, but in theory it could also AV the process. Not something we can test for unfortunately.

File: src/libraries/System.Private.CoreLib/src/System/Array.cs (hunk 11 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/Array.cs]: This PR removes the internal `GetRawSzArrayData` method in favor of the new public `MemoryMarshal` API. This particular call site has a bit of a smell unfortunately since the input isn't already of type `T[]`, but the majority of the call sites throughout the PR do look cleaner. See below for a canonical example of how most call sites no longer need `Unsafe.As` with this new API.

https://github.com/dotnet/runtime/pull/1036/files#diff-d1745c212e8b7f4a46bc702c8d511cadR41

**Comment by GrabYourPitchforks** [comment]
Derp - I forgot the libraries part of the change.

**Comment by GrabYourPitchforks** [comment]
Test failures were caused by https://github.com/dotnet/runtime/issues/1386, which is now fixed. Merging this PR.

**Comment by MichalStrehovsky** [comment]
There's another reference to the deleted API in crossgen2 here: https://github.com/dotnet/runtime/blob/c4a7c7f9365879addd573369cb49995aa4c61fef/src/coreclr/src/tools/Common/TypeSystem/Interop/IL/Marshaller.cs#L1325

I'm on my phone now, done for the day (and still on parental leave anyway). In case someone hits this before it's fixed, cc @dotnet/crossgen-contrib

**Comment by GrabYourPitchforks** [comment]
@MichalStrehovsky Thanks for the heads up - will get a new PR out for those.
